File: heapsort.c
#include<stdio.h>
// swap function
void swap(int *a, int *b){
	int temp;
	temp = *a;
	*a = *b;
	*b = temp;
}

void heapify(int arr[], int n, int i){
	
	// find largest among root, left child and right child
	int largest = i; // root
	int left = 2 * i + 1;
	int right = 2 * i + 2;

	if (left < n && arr[left] > arr[largest])
		largest = left;

	if (right < n && arr[right] > arr[largest])
		largest = right; 

	// swap continiues if root is not largest 

	if (largest != i) {
	swap (&arr[i], &arr[largest]);
	heapify (arr, n, largest);
	}
}
// do heapsort

void heapSort(int arr[], int n){

	// building max heap
	for (int i = (n / 2) - 1; i >= 0; i--)
		heapify(arr, n, i);

	// HeapSort
	for (int i = n -1; i >= 0; i--){
		swap(&arr[0], &arr[i]);
		
		// heapify root element to get highest element at root again 
		heapify(arr, i, 0);
	}
}

// print array 

void printArray(int arr[], int n){
	for (int i = 0; i < n; ++i)
		printf("%d ", arr[i]);
	printf("\n");
}

// main function 

int main(){
	int arr[] = {1, 2, 24, 46, 57, 88, 102};
	int n = sizeof(arr) / sizeof(arr[0]);

	heapSort(arr, n);

	printf("Sorted array\n");
	printArray(arr, n);	
	return 0;
}

File: heapsort.py
def heapify(arr, n, i):
	
	# largest among root and child	

	largest = i
	left = 2 * i + 1
	right = 2 * i + 2

	if left < n and arr[i] < arr[left]:
		largest = left

	if right < n and arr[i] < arr [right]:
		largest = right
	
	# if root is not largest, swap with largest and continue heapifying
	
	if largest != i:
		arr[i], arr[largest] = arr[largest], arr[i]
		heapify(arr, n, largest)

def heapSort(arr):
	n = len(arr)

	# build max heap

	for i in range(n//2, -1, -1):
		heapify(arr, n, i)

	for i in range(n-1, 0, n-1):
		# swap
		arr[i], arr[0] = arr[0], arr[i]

		# hepify root element
		heapify(arr, i, 0)

arr = [1, 2, 24, 46, 57, 88, 102]
heapSort(arr)
n = len(arr)
print("Sorted array")
for i in range (n):
	print("%d " % arr[i], end = '')
